File: php/src/Webhooks/SignatureVerifier.php
<?php

declare(strict_types=1);

namespace Spart\Sdk\Webhooks;

use Spart\Sdk\Exceptions\SpartValidationException;

/**
 * Verifies inbound Spart webhook signatures.
 *
 * Header format: `X-Spart-Signature: t=<unix-seconds>,v1=<lowercase-hex>`
 * Signature: HMAC-SHA256 over the byte sequence `"{t}.{rawBody}"`.
 *
 * @final
 */
final class SignatureVerifier
{
    public function __construct(
        private readonly string $signingSecret,
        private readonly int $toleranceSeconds = 300,
    ) {
        if (trim($this->signingSecret) === '') {
            throw new \InvalidArgumentException('SignatureVerifier::signingSecret must not be blank.');
        }
        if ($this->toleranceSeconds < 1 || $this->toleranceSeconds > 86400) {
            throw new \InvalidArgumentException(
                'SignatureVerifier::toleranceSeconds must be between 1 and 86400 seconds (24h).'
            );
        }
    }

    /**
     * @param int|null $now Unix seconds used for the timestamp-tolerance check;
     *                      defaults to the wall clock. Injectable for deterministic tests.
     */
    public function verify(string $rawBody, string $headerValue, ?int $now = null): bool
    {
        $parsed = self::parseHeader($headerValue);
        if ($parsed === null) {
            return false;
        }
        [$t, $v1] = $parsed;

        $now ??= time();
        if (abs($now - $t) > $this->toleranceSeconds) {
            return false;
        }

        $expected = hash_hmac('sha256', "{$t}.{$rawBody}", $this->signingSecret);
        return hash_equals($expected, $v1);
    }

    /**
     * @throws SpartValidationException when the signature does not verify
     *                                  or the envelope cannot be parsed.
     */
    public function verifyAndParse(string $rawBody, string $headerValue, string $deliveryId, int $attempt): Event
    {
        if (!$this->verify($rawBody, $headerValue)) {
            throw new SpartValidationException('Spart webhook signature verification failed.');
        }
        try {
            return Event::fromJson($rawBody, deliveryId: $deliveryId, attempt: $attempt);
        } catch (\InvalidArgumentException $e) {
            throw new SpartValidationException(
                'Spart webhook envelope is invalid: ' . $e->getMessage(),
                previous: $e,
            );
        }
    }

    /** @return array{0:int,1:string}|null */
    public static function parseHeader(string $value): ?array
    {
        if ($value === '') {
            return null;
        }
        $t = null;
        $v1 = null;
        foreach (explode(',', $value) as $part) {
            $part = trim($part);
            if (str_starts_with($part, 't=')) {
                $raw = substr($part, 2);
                if (!ctype_digit($raw)) {
                    return null;
                }
                $t = (int) $raw;
            } elseif (str_starts_with($part, 'v1=')) {
                $v1 = substr($part, 3);
            }
        }
        if ($t === null || $v1 === null || $v1 === '') {
            return null;
        }
        return [$t, $v1];
    }
}

File: python/src/spart_sdk/signature_verifier.py
from __future__ import annotations

import hashlib
import hmac
import re
import time

_DEFAULT_TOLERANCE_SECONDS = 300
_MAX_TOLERANCE_SECONDS = 86_400
_DIGITS = re.compile(r"^[0-9]+$")


class SignatureVerifier:
    """Verifies inbound Spart webhook signatures.

    Header format: ``X-Spart-Signature: t=<unix-seconds>,v1=<lowercase-hex>``.
    Signature: HMAC-SHA256 over the bytes ``f"{t}.".encode() + raw_body``.
    """

    def __init__(self, secret: str, tolerance_seconds: int = _DEFAULT_TOLERANCE_SECONDS) -> None:
        if secret.strip() == "":
            raise ValueError("SignatureVerifier: secret must not be blank.")
        if not 1 <= tolerance_seconds <= _MAX_TOLERANCE_SECONDS:
            raise ValueError("SignatureVerifier: tolerance_seconds must be between 1 and 86400.")
        self._secret = secret.encode("utf-8")
        self._tolerance_seconds = tolerance_seconds

    def verify(self, raw_body: bytes, signature_header: str, now: int | None = None) -> bool:
        parsed = self.parse_header(signature_header)
        if parsed is None:
            return False
        t, v1 = parsed

        current = int(time.time()) if now is None else now
        if abs(current - t) > self._tolerance_seconds:
            return False

        signed = f"{t}.".encode("utf-8") + raw_body
        expected = hmac.new(self._secret, signed, hashlib.sha256).hexdigest()
        # Compare on bytes: hmac.compare_digest raises on non-ASCII str operands,
        # whereas v1 is attacker-controlled. Encoding lets a non-hex v1 fail
        # closed (return False) instead of raising, matching the other SDKs.
        return hmac.compare_digest(expected.encode("ascii"), v1.encode("utf-8"))

    @staticmethod
    def parse_header(value: str) -> tuple[int, str] | None:
        if value == "":
            return None
        t: int | None = None
        v1: str | None = None
        for part in value.split(","):
            part = part.strip()
            if part.startswith("t="):
                raw = part[2:]
                if not _DIGITS.match(raw):
                    return None
                t = int(raw)
            elif part.startswith("v1="):
                v1 = part[3:]
        if t is None or v1 is None or v1 == "":
            return None
        return t, v1
